### PR TITLE
feat(client,api): add user tokens and route for submitting coderoad tutorials

### PR DESCRIPTION
**Pull request by moT01**
This adds the first part needed to be able to submit tutorials through CodeRoad. It adds tokens for users that they can create and delete. It also adds a few variables for cert claiming on the back end that will be needed at some point. 

Edit: This also adds the second part - the route for submitting the tutorials

~~I am going to open this up for review, but continue working on adding the rest of what's needed - I will probably turn that into a separate PR. I haven't worked with some of these areas of the code base much - particularly, on the back end. It might be a good idea to look a little closer at that area. The front end should be pretty solid. All advice and thoughts are welcome.~~

~~The widget only shows up on the relational databases superblock page, so you need to have the upcoming changes visible to see it.~~

Edit: There are two widgets, a mini one on the relational databases superBlock page - and a second one on the settings page. You need to have the upcoming changes visible to see it.

- [x] I have read [freeCodeCamp's contribution guidelines](https://contribute.freecodecamp.org).
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`)
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Related to: https://github.com/freeCodeCamp/freeCodeCamp/issues/42509

### CODE REVIEW

**Review by ShaunSHamilton — DISMISSED** [review]
Quick small suggestion... I am happy to go with `deleet` if that is what you wanted, though.

File: client/i18n/locales/english/translations.json
Comment (nhcarrigan) [inline client/i18n/locales/english/translations.json:606]:
```suggestion
    "create-p1": "It looks like you don't have a webhook token. Create one to save your progress on this section",
    "create-p2": "Create a webhook token to save your progress on the curriculum sections that use a virtual machine.",
    "delete": "Delete my token",
    "delete-title": "Delete My Webhook Token",
    "delete-p1": "Your webhook token below is used to save your progress on the curriculum sections that use a virtual machine.",
    "delete-p2": "If you suspect your token has been compromised, you can delete it to make it unusable. Progress on previously submitted lessons will not be lost.",
    "delete-p3": "You will need to create a new token to save future progress on the curriculum sections that use a virtual machine.",
```

Comment (moT01) [reply client/i18n/locales/english/translations.json]: Just to clarify, this change doesn't need to be added to the rest of the languages, correct?

**Comment by moT01** [comment]
Alright, I think we can probably review and merge this - I updated the PR description.

**Comment by moT01** [comment]
I think this is ready for review again. It shouldn't change any visible, existing functionality on the site.

There' two parts to this: 
- the token generation for users
- and the route for submitting challenges

The token generation:
Users that haven't created a token will see a box on on the RDBMS superBlock page (superBlock still hidden via `upcomingChange`) saying that they "need a token to save their progress" or something. There's a button there to create one. This will add a `WebhookToken` collection to the database with a new token and the user Id. There will be a similar box on the settings page (hidden via comments in the PR) giving them an option to delete their token or create one if they don't have one.

The route:
This token will be passed to CodeAlly, and used by CodeRoad - when a tutorial is complete, coderoad will hit the endpoint I made, `/coderoad-challenge-completed` - defined in each [tutorial repo](https://github.com/moT01/demo-fcc-cert-project/blob/main/coderoad.yaml#L18). The request will include the user token and the repo name - there's an example in the PR. Then, after some validation, the tutorial will be saved to the user's completed challenges.

Some things to maybe look a little closer at in this PR:
I opened the `coderoad-challenge-completed` route to accept requests without an access or csrf token. I don't see a problem with that, but I'm not the security expert and I don't see another way to do it.
Database updates - I did some pretty thorough testing, it looks like it all works - but I would hate to cause problems with database entries, so maybe just look closely there.

This was not easy to test for me so if anyone wants to get together for an explanation or demo on how it works, let me know. I plan on creating a group of repos that can be used for testing.

**Review by ShaunSHamilton — DISMISSED** [review]
Minor thing, and some queries.

Otherwise (not necessary for this PR), after clicking the _Create Token_ button on the _Relational DB_ superblock, I was confused why it did not tell me anything.

Oddly, I expected either to be told what my token was. Or, at least, _"Your webhook token has been created and successfully saved"_

**Comment by moT01** [comment]
> Is the coderoad bit supposed to be working yet?

Well - ideally, it would be working. Still some bugs. I tried to load it up today and saw that - made an issue: `https://github.com/freeCodeCamp/CodeAlly-CodeRoad-freeCodeCamp/issues/28` Doesn't affect this PR for now.

**Review by ShaunSHamilton — DISMISSED** [review]

**Review by ShaunSHamilton — DISMISSED** [review]
Still good to go? 👍

**Review by ShaunSHamilton — DISMISSED** [review]

**Review by raisedadead — COMMENTED** [review]
Hey @moT01 - You mentioned in the chat, this is to go out this week. I see some commented out code (boolean flags), etc. Is this ready to go?

File: client/src/templates/Challenges/codeally/show.tsx
Comment (ojeytonwilliams) [inline client/src/templates/Challenges/codeally/show.tsx:82]: Sending something that should be private via query parameter is [a bit sketchy](https://owasp.org/www-community/vulnerabilities/Information_exposure_through_query_strings_in_url). Is there another way we can send the token to CodeAlly?

File: client/src/redux/index.js
Comment (ojeytonwilliams) [inline client/src/redux/index.js:653]:
```suggestion
```
This seems redundant.

**Comment by ojeytonwilliams** [comment]
I've been slowly reviewing this and there are a few things I'd like to discuss before we get it in - I'll finish the review ASAP.  Sorry for the delay.

File: client/src/redux/index.js
Comment (ojeytonwilliams) [inline client/src/redux/index.js:669]:
```suggestion
```
Again, seems redundant.

File: client/src/redux/index.js
Comment (ojeytonwilliams) [inline client/src/redux/index.js:678]:
```suggestion
            webhookToken: ''
```
Or null, perhaps?  I'm not quite sure why we'd need the payload, since we know that it's been deleted.

File: client/src/utils/ajax.ts
Comment (ojeytonwilliams) [inline client/src/utils/ajax.ts:217]:
```suggestion
  return post('/user/webhook-token');
```
Is the empty body necessary here?

File: client/src/utils/ajax.ts
Comment (ojeytonwilliams) [inline client/src/utils/ajax.ts:265]:
```suggestion
  return deleteRequest('/user/webhook-token');
```
Again, seem unnecessary.

**Review by ojeytonwilliams — CHANGES_REQUESTED** [review]
One thing I don't quite get is, what am I expecting to see CodeAlly?  I see that the token is passed as part of the iframe embed, but I'm not sure exactly what CodeAlly does with it.

**Comment by ojeytonwilliams** [comment]
Btw, sending the token via query param is probably fine, assuming you can't do anything destructive with it.

File: api-server/src/server/boot/challenge.js
Comment (moT01) [reply api-server/src/server/boot/challenge.js:413]: The request comes from coderoad, that's where the response goes. There's nothing they can do with that info. Basically, this route will just add it to our database. Their heatmap and completed challenges etc will be updated the next time they refresh the page. I didn't test it out - I'm hoping that when a user comes out of the VM, it will be updated - but, thinking about it, they will probably need a hard refresh.

File: client/src/templates/Challenges/codeally/show.tsx
Comment (moT01) [reply client/src/templates/Challenges/codeally/show.tsx:82]: You're probably right. There's not another way at the moment. Another plan we had at one point was to have users manually input their token within coderoad. It was decided to go this route. Unless there's some significant issue with it, I think this will have to work.

**Comment by moT01** [comment]
> what am I expecting to see CodeAlly?

CodeAlly adds it to the env variables. You should be able to see it using `printenv` in the terminal of the VM. CodeRoad must have access to those.

> assuming you can't do anything destructive with it.

The only thing you should be able to do at the moment is submit challenges with it. I am thinking that we can potentially use it for submitting other challenges - like the data science stuff. Nothing destructive anyway.

File: client/src/utils/ajax.ts
Comment (moT01) [reply client/src/utils/ajax.ts:217]: Currently, they sort of are - the helper functions at the top of this file expect two arguments. We could make the second one (body) optional. There were several other uses of these functions that used an empty object as the second argument, so I did what those were doing. Any preference?

Comment (ojeytonwilliams) [reply client/src/utils/ajax.ts:217]: I think, since there's no need to add a body just because you're POSTing, that we should make it optional.  I'll create a note for myself to change this after we merge, though, what you've done seems good for this PR.

**Comment by ojeytonwilliams** [comment]
> > what am I expecting to see CodeAlly?
> 
> CodeAlly adds it to the env variables. You should be able to see it using printenv in the terminal of the VM. CodeRoad must have access to those.

Great, thanks.  It's not doing anything with them, yet, right?

**Comment by moT01** [comment]
> It's not doing anything with them...

It's doing what it needs to do with them. It adds them to the environment for CodeRoad to access. CodeAlly needs to update CodeRoad so it uses the new version with the webhook functionality. Basically, CodeRoad has something like: `if (token) -> send request when asked`. Then 'when asked' part is something I need to add to each tutorial, there's an [example here](https://github.com/moT01/demo-fcc-cert-project/blob/main/coderoad.yaml#L17)

**Comment by moT01** [comment]
Should be good to go 👍

**Review by raisedadead — COMMENTED** [review]
Thanks for the quick turnaround on this folks. I would want to push this out as a dedicated release once this is ready to be merged into main.

**Comment by moT01** [comment]
What exactly does 'dedicated release' mean? We deploy, then merge only this and deploy again? Just to be clear, this doesn't actually add the curriculum to the site yet, it's all still hidden.

**Review by ojeytonwilliams — APPROVED** [review]
LGTM :+1: 

Sorry for the drawn out review, but this looks great.  Nice work, @moT01

**Review by ShaunSHamilton — APPROVED** [review]
This LGTM 👍 . Nice one getting this done, Tom

**Comment by ShaunSHamilton** [comment]
@raisedadead Do you want to hold off merging this until we are ready to dedicate a release to it? Or, will we just cherry pick it when the time comes?

Either way, I leave it to you to merge.

**Comment by ojeytonwilliams** [comment]
I might be missing something, but I think this is safe to deploy as-is.  The new features are either new endpoints or hidden behind the feature flag, so they should be invisible to users.

**Comment by moT01** [comment]
Thanks again for all the review everyone 🎉 This was a big one 👍

**Pull request merged**
